File: tests/evals/multi-file/untangle-the-invoice-tax-cycle/solution/billing/invoice.py
"""Invoices."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from billing.money import normalize_currency, to_minor_units
from billing.tax import tax_for_line


@dataclass(frozen=True)
class Line:
    description: str
    unit_amount: float
    quantity: int = 1
    category: str = "default"


@dataclass(frozen=True)
class Invoice:
    currency: str
    lines: tuple[Line, ...] = ()

    def totals(self) -> dict[str, Any]:
        """Subtotal, tax and total, all in integer minor units."""
        currency = normalize_currency(self.currency)
        subtotal = 0
        tax = 0
        for line in self.lines:
            minor = to_minor_units(line.unit_amount, currency) * line.quantity
            subtotal += minor
            tax += tax_for_line(minor, currency, line.category)
        return {
            "currency": currency,
            "subtotal_minor": subtotal,
            "tax_minor": tax,
            "total_minor": subtotal + tax,
        }

File: tests/evals/multi-file/rename-a-barrel-exported-formatter/fixture/src/invoice.js
'use strict';

const { formatAmount, formatPercent } = require('./format');

/**
 * A receipt, one line per item plus the totals block.
 */
function renderInvoice(invoice) {
  const lines = invoice.lines.map(
    (line) => `${line.quantity} x ${line.description}  ${formatAmount(line.unitMinor, invoice.currency)}`,
  );
  lines.push(`subtotal ${formatAmount(invoice.subtotalMinor, invoice.currency)}`);
  lines.push(`tax ${formatAmount(invoice.taxMinor, invoice.currency)} (${formatPercent(invoice.taxRateBp)})`);
  lines.push(`total ${formatAmount(invoice.subtotalMinor + invoice.taxMinor, invoice.currency)}`);
  return lines.join('\n');
}

module.exports = { renderInvoice };
